program/src/instruction.rs

```diff
@@ -1,184 +1,420 @@
+    )]
+    #[account(
+        2,
+        name = "config_authority",
+        desc = "Config authority"
+    )]
+    #[account(
+        3,
+        name = "mint",
+        desc = "Stake token mint"
+    )]
+    #[account(
+        4,
+        name = "vault",
+        desc = "Stake token vault"
+    )]
     InitializeConfig {
-        cooldown_time: UnixTimestamp,
+        cooldown_time_seconds: UnixTimestamp,
         max_deactivation_basis_points: u16,
     },
 
     /// Initializes stake account data for a validator.
-    ///
+    /// 
     /// NOTE: Anybody can create the stake account for a validator. For new
     /// accounts, the authority is initialized to the validator vote account's
     /// withdraw authority.
-    ///
-    /// 0. `[]` Stake config account
-    /// 1. `[w]` Validator stake account
-    ///     * PDA seeds: ['stake', validator, config_account]
-    /// 2. `[]` Validator vote account
-    /// 3. `[]` System program
+    #[account(
+        0,
+        name = "config",
+        desc = "Stake config account"
+    )]
+    #[account(
+        1,
+        writable,
+        name = "stake",
+        desc = "Validator stake account (pda of `['stake', validator, config]`)"
+    )]
+    #[account(
+        2,
+        name = "validator_vote",
+        desc = "Validator vote account"
+    )]
+    #[account(
+        3,
+        name = "system_program",
+        desc = "System program account"
+    )]
     InitializeStake,
 
     /// Stakes tokens with the given config.
-    ///
+    /// 
     /// Limited to the current amount of SOL staked to the validator.
     ///
     /// NOTE: Anybody can stake tokens to a validator, but this does not work
     /// like native staking, because the validator can take control of staked
     /// tokens by deactivating and withdrawing.
-    ///
-    /// 0. `[w]` Stake config account
-    /// 1. `[w]` Validator stake account
-    ///     * PDA seeds: ['stake', validator, config_account]
-    /// 2. `[w]` Token Account
-    /// 3. `[s]` Owner or delegate of the token account
-    /// 4. `[]` Validator vote account
-    /// 3. `[]` Stake Token Mint
-    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
-    ///   Must be the token account on the stake config account
-    /// 5. `[]` Token program
-    /// 6.. Extra accounts required for the transfer hook
-    ///
-    /// Instruction data: amount of tokens to stake, as a little-endian u64
+    /// 
+    /// Instruction data: amount of tokens to stake, as a little-endian `u64`.
+    #[account(
+        0,
+        writable,
+        name = "config",
+        desc = "Stake config account"
+    )]
+    #[account(
+        1,
+        writable,
+        name = "stake",
+        desc = "Validator stake account"
+    )]
+    #[account(
+        2,
+        writable,
+        name = "source_token",
+        desc = "Token account"
+    )]
+    #[account(
+        3,
+        signer,
+        name = "token_authority",
```

Comment:
nit: can this be `token_account_authority`? Back in 2020, I tried to call this the `transfer_authority`, but it never caught on :sweat_smile:

Reply:
Done.